### PR TITLE
[FLASK] Add snap alerts and prompts via `snap_dialog` RPC method

### LINKED ISSUE #613: Expand snap confirmation capabilities
(This may be irrelevant if we implement a blockkit like UI)

In addition to the confirmation via `snap_confirm`, we need:
- A prompt with plaintext input (yes/no)
- An alert (not yes/no, just "ok")

All of these things could be encompassed within the same permission, and ideally they should. Maybe something like `snap_dialog` or `_createDialog`, with a `type` enum. The alert type could always return `true`, and the prompt / confirmation would return `true`/`false` depending on the user's choice.

### PR DESCRIPTION
**Pull request by rekmarks**
Snaps can already create confirmation dialogs, i.e. dialogs with a message and yes/no buttons, via the `snap_confirm` RPC method. This PR adds support for a new RPC method, `snap_dialog`, that enables snaps to also create:
- alerts, i.e. dialogs with a message and a single "OK" button
- prompts, i.e. dialogs with a text input field and cancel/submit buttons

Fixes https://github.com/MetaMask/snaps-monorepo/issues/613

### CODE REVIEW

File: ui/pages/confirmation/confirmation.js [hunk 1]
```diff
@@ -21,13 +22,16 @@ import {
 } from '../../helpers/constants/design-system';
 import { useI18nContext } from '../../hooks/useI18nContext';
 import { useOriginMetadata } from '../../hooks/useOriginMetadata';
-import { getUnapprovedTemplatedConfirmations } from '../../selectors';
+import { getSnaps, getUnapprovedTemplatedConfirmations } from '../../selectors';
```
Comment (FrederikBolding) [inline ui/pages/confirmation/confirmation.js:25]: This should be fenced

**Comment by FrederikBolding** [comment]
I've asked @eriknson to review the design of this. I'll do a code review. I have also verified that all the dialogs are working as intended.

File: ui/pages/confirmation/confirmation.js [hunk 2]
```diff
@@ -132,14 +141,66 @@ export default function ConfirmationPage({
...
+    t,
+    dispatch,
+    history,
+    ///: BEGIN:ONLY_INCLUDE_IN(flask)
+    isSnapDialog,
+    snaps,
+    ///: END:ONLY_INCLUDE_IN
+  ]);
+
+  const INPUT_STATE_CONFIRMATIONS = [MESSAGE_TYPE.SNAP_DIALOG_PROMPT];
```
Comment (FrederikBolding) [inline ui/pages/confirmation/confirmation.js:192]: `MESSAGE_TYPE.SNAP_DIALOG_PROMPT` should be fenced

File: ui/pages/confirmation/confirmation.js [hunk 3]
```diff
@@ -147,7 +147,9 @@ export default function ConfirmationPage({
   };
 
   ///: BEGIN:ONLY_INCLUDE_IN(flask)
-  const snaps = useSelector(getSnaps);
+  const {
+    manifest: { snapName: proposedName },
+  } = useSelector(getSnap);
```
Comment (FrederikBolding) [inline ui/pages/confirmation/confirmation.js:152]: You need to pass the snapId to the selector

Comment (GuillaumeRx) [reply ui/pages/confirmation/confirmation.js:152]: Oh yeah 😅

Comment (FrederikBolding) [reply ui/pages/confirmation/confirmation.js:152]: https://github.com/MetaMask/metamask-extension/blob/4082c33eb62a2411512e54c619a4504e70283ac0/ui/components/app/flask/snap-ui-renderer/snap-ui-renderer.js#L62

**Review by FrederikBolding — APPROVED** [review]

**Review by FrederikBolding — DISMISSED** [review]

File: ui/components/app/flask/copyable/copyable.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      paddingTop={2}
+      paddingBottom={2}
+    >
+      <Typography
+        color={COLORS.TEXT_ALTERNATIVE}
+        marginRight={2}
+        overflowWrap={OVERFLOW_WRAP.ANYWHERE}
+      >
+        {text}
+      </Typography>
```
Comment (eriknson) [inline ui/components/app/flask/copyable/copyable.js:36]: Let's make text inside the `<Copyable />` container same font-size (_Body-MD_) as the body text outside of it; `16px` or `0.875rem`

File: ui/components/app/flask/snap-delineator/snap-delineator.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        <Icon name="snaps-filled" color={COLORS.INFO_DEFAULT} size={SIZES.SM} />
+        <Text
+          variant={TEXT.BODY_SM}
+          color={COLORS.INFO_DEFAULT}
+          className="snap-delineator__header__text"
+          marginLeft={1}
+          marginTop={0}
+          marginBottom={0}
+        >
+          {t('contentFromSnap', [snapName])}
```
Comment (eriknson) [inline ui/components/app/flask/snap-delineator/snap-delineator.js:43]: After _Content from_ the `snapName` is not rendering right now.

Screenshot 2022-12-01 at 11 41 20

File: ui/components/app/flask/copyable/copyable.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        marginBottom={1}
+      >
+        {copied ? (
+          <IconCopied color="var(--color-icon-alternative)" size={18} />
+        ) : (
+          <IconCopy
+            color="var(--color-icon-alternative)"
+            onClick={() => handleCopy(text)}
+            size={18}
+          />
```
Comment (eriknson) [inline ui/components/app/flask/copyable/copyable.js:51]: Let's add `cursor: pointer` to when it's clickable / copyable.

File: ui/pages/confirmation/components/confirmation-footer/confirmation-footer.js [hunk 7]
```diff
@@ -3,21 +3,23 @@ import PropTypes from 'prop-types';
...
-        <Button type="primary" onClick={onApprove}>
-          {approveText}
+        {onCancel ? (
+          <Button type="secondary" onClick={onCancel}>
+            {cancelText}
+          </Button>
+        ) : null}
+        <Button type="primary" onClick={onSubmit}>
+          {submitText}
         </Button>
```
Comment (eriknson) [inline ui/pages/confirmation/components/confirmation-footer/confirmation-footer.js:23]: When there's a single button in the confirmation footer, like in the alert-case, let's remove the `margin-right: 16px `

**Review by eriknson — DISMISSED** [review]
Design review

File: ui/pages/confirmation/confirmation.js [hunk 8]
```diff
@@ -132,14 +141,70 @@ export default function ConfirmationPage({
...
   const [alertState, dismissAlert] = useAlertState(pendingConfirmation);
 
+  const [inputStates, setInputStates] = useState({});
+  const setInputState = (key, value) => {
+    setInputStates((currentState) => ({ ...currentState, [key]: value }));
+  };
+
+  ///: BEGIN:ONLY_INCLUDE_IN(flask)
+  const {
+    manifest: { snapName: proposedName },
```
Comment (FrederikBolding) [inline ui/pages/confirmation/confirmation.js:151]:
```suggestion
    manifest: { proposedName },
```

Comment (FrederikBolding) [reply ui/pages/confirmation/confirmation.js:151]: Should fix the error that @eriknson mentioned above

File: ui/components/app/flask/snap-delineator/snap-delineator.js (hunk 5 above)
Comment (FrederikBolding) [reply ui/components/app/flask/snap-delineator/snap-delineator.js:43]: This worked previously when I tested, but must have missed it breaking when I suggested some changes. Left a suggestion to fix below.

**Review by FrederikBolding — DISMISSED** [review]

**Review by FrederikBolding — APPROVED** [review]

**Review by hmalik88 — APPROVED** [review]
LGTM

**Pull request merged**